File: 0069-sqrtx/0069-sqrtx.java
class Solution {
    public int mySqrt(int x) {
        if (x < 2) return x;

        int low = 1;
        int high = x / 2; // Optimization: sqrt(x) <= x / 2 for x >= 4

        while (low <= high) {
            int mid = low + (high - low) / 2;
            long square = (long) mid * mid; // Avoid overflow

            if (square == x) {
                return mid;
            } else if (square < x) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return high; // high will be the integer part of sqrt(x)
    }
}

File: 0069-sqrtx/0069-sqrtx.py
class Solution(object):
    def mySqrt(self, x):
        """
        :type x: int
        :rtype: int
        """
        if x < 2:
            return x

        low = 1
        high = x // 2  # Optimization

        while low <= high:
            mid = low + (high - low) // 2
            square = mid * mid

            if square == x:
                return mid
            elif square < x:
                low = mid + 1
            else:
                high = mid - 1  # Corrected from your Java version

        return high  # high is the integer part of sqrt(x)
